src/sources/mongodb_metrics/mod.rs

```diff
@@ -0,0 +1,1065 @@
+    CommandIsMasterMongoError { source: MongoError },
+    #[snafu(display("failed to parse `isMaster` response: {:?}", source))]
+    CommandIsMasterParseError { source: bson::de::Error },
+    #[snafu(display("failed to execute `buildInfo` command: {:?}", source))]
+    CommandBuildInfoMongoError { source: MongoError },
+    #[snafu(display("failed to parse `buildInfo` response: {:?}", source))]
+    CommandBuildInfoParseError { source: bson::de::Error },
+    #[snafu(display("only `Mongod` supported right now, current: `{:?}`", node_type))]
+    UnsupportedNodeType { node_type: NodeType },
+}
+
+#[derive(Debug)]
+enum CollectError {
+    Mongo(MongoError),
+    Bson(bson::de::Error),
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, Default)]
+#[serde(deny_unknown_fields)]
+struct MongoDBMetricsConfig {
+    endpoint: String,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+    #[serde(default = "default_namespace")]
+    namespace: String,
+}
+
+#[derive(Debug)]
+struct MongoDBMetrics {
+    client: Client,
+    endpoint: String,
+    namespace: String,
+    tags: BTreeMap<String, String>,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+pub fn default_namespace() -> String {
+    "mongodb".to_string()
+}
+
+inventory::submit! {
+    SourceDescription::new::<MongoDBMetricsConfig>("mongodb_metrics")
+}
+
+impl_generate_config_from_default!(MongoDBMetricsConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "mongodb_metrics")]
+impl SourceConfig for MongoDBMetricsConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        let mongodb = MongoDBMetrics::new(&self.endpoint, &self.namespace).await?;
+
+        let out = out
+            .sink_map_err(|e| error!("error sending metric: {:?}", e))
+            .sink_compat();
+
+        let task = interval(Duration::from_secs(self.scrape_interval_secs))
+            .take_until(shutdown.compat())
+            .filter_map(move |_| {
+                let mongodb = Arc::clone(&mongodb);
+                async { Some(mongodb.collect().await) }
+            })
+            .flatten()
+            .map(Event::Metric)
+            .map(Ok)
+            .forward(out)
+            .inspect(|_| info!("finished sending"));
+
+        Ok(Box::new(task.boxed().compat()))
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "mongodb_metrics"
+    }
+}
+
+impl MongoDBMetrics {
+    /// Works only with Standalone connection-string. Collect metrics only from specified instance.
+    /// https://docs.mongodb.com/manual/reference/connection-string/#standard-connection-string-format
+    async fn new(endpoint: &str, namespace: &str) -> Result<Arc<MongoDBMetrics>, BuildError> {
```

Comment:
Could you explain a little why we are using `Arc` here and below? I'm not super familiar with `Arc` but I was a bit surprised to see it pop up in this context.

Reply:
[Arc](https://doc.rust-lang.org/std/sync/struct.Arc.html) is a thread-safe version of [Rc](https://doc.rust-lang.org/std/rc/struct.Rc.html) (`Arc` implement [Send](https://doc.rust-lang.org/std/marker/trait.Send.html) and [Sync](https://doc.rust-lang.org/std/marker/trait.Sync.html), while `Rc` is not). I call `async` function `MongoDBMetrics::collect` in each item in `interval` stream. Because `collect` is async I need `MongoDBMetrics` instance for each `Future` (if we want `'static` lifetime). There few solutions: 1) Create new object each time 2) Clone object (if clone supported) 3) Use `Arc` as cheaper clone (might be not cheaper for simple structs).
Looking on the code again, I think I'd able to collect metrics without `Arc` (what should be simpler). Will try to change with `endpoint` => `endpoints`.

Reply:
Arc removed.

Reply:
:+1: thanks for the explanation too, that is educational.